### PR TITLE
green node endpoint replacement

### PR DESCRIPTION
**Pull request by sergiyvamz**
### Summary

green node endpoint replacement

### Description

Allows the driver to to replace `Green` DNS endpoints when they get unavailable after Blue/Green switchover is completed.

### Additional Reviewers

@karenc-bq 
@davecramer 

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/using-the-jdbc-driver/using-plugins/UsingTheFailoverPlugin.md [hunk 1]
```diff
@@ -104,3 +104,8 @@ public class HikariCPSQLException implements SQLExceptionOverride {
     }
 }
 ```
+
+### Blue/Green Deployments
+Although AWS Advanced JDBC Driver is not compatible with [AWS Blue/Green Deployments](https://docs.aws.amazon.com/whitepapers/latest/overview-deployment-options/bluegreen-deployments.html) and officially doesn't support it, the Failover Plugin has been validated for clusters with Blue/Green Deployments. While general basic connectivity to `Blue` and `Green` clusters
+are always in place, some failover cases are not fully supported. The team is aware of these cases and working on them to bring
+AWS Blue/Green Deployments support at full scale. Meanwhile, the users may consider using configuration parameter [`enableGreenNodeReplacement`](../UsingTheJdbcDriver.md) that allows the driver to replace `Green` DNS endpoints when they get unavailable after Blue/Green switchover is completed.
```
Comment (davecramer) [inline docs/using-the-jdbc-driver/using-plugins/UsingTheFailoverPlugin.md:111]: s/when they get unavailable'/'when they become unavailable'

File: wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java [hunk 2]
```diff
@@ -117,12 +122,69 @@ public Connection connect(
...
+      final String fixedHost = this.rdsUtils.removeGreenInstancePrefix(hostSpec.getHost());
+      final HostSpec connectionHostSpec = new HostSpecBuilder(hostSpec.getHostAvailabilityStrategy())
+          .copyFrom(hostSpec)
+          .host(fixedHost)
+          .build();
+
+      final ConnectInfo fixedConnectInfo = targetDriverDialect.prepareConnectInfo(protocol, connectionHostSpec, copy);
+
+      LOGGER.finest(() -> "Connecting (with fixed host) to " + fixedConnectInfo.url
+          + PropertyUtils.logProperties(PropertyUtils.maskProperties(fixedConnectInfo.props), "\nwith properties: \n"));
```
Comment (davecramer) [inline wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java:184]: This logger message won't make much sense outside of the context of this code.
Something like "Connecting to" + fixedConnectionInfo.url + " after correcting the hostname from" + oldHostName

File: wrapper/src/main/java/software/amazon/jdbc/PropertyDefinition.java [hunk 3]
```diff
@@ -156,6 +156,16 @@ public class PropertyDefinition {
               "true", "false"
           });
 
+  public static final AwsWrapperProperty ENABLE_GREEN_NODE_REPLACEMENT =
+      new AwsWrapperProperty(
+          "enableGreenNodeReplacement",
+          "false",
+          "Enables to replace a green node host names with original host name when green host DNS doesn't exist.",
```
Comment (davecramer) [inline wrapper/src/main/java/software/amazon/jdbc/PropertyDefinition.java:163]: Enables replacing a green node host name with the original hostname after a blue/green switchover and the green name no longer resolves

**Review by davecramer — APPROVED** [review]

File: wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java [hunk 4]
```diff
@@ -157,6 +163,73 @@ public Connection connect(
...
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props)
+      throws SQLException {
+
+    try {
+      return ds.getConnection();
+
+    } catch (Throwable throwable) {
+
+      if (!PropertyDefinition.ENABLE_GREEN_NODE_REPLACEMENT.getBoolean(props)) {
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java:179]:
```suggestion
      return ds.getConnection();
    } catch (Throwable throwable) {
      if (!PropertyDefinition.ENABLE_GREEN_NODE_REPLACEMENT.getBoolean(props)) {
```

File: wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java [hunk 5]
```diff
@@ -157,6 +163,73 @@ public Connection connect(
...
+        loopThrowable = loopThrowable.getCause();
+      }
+
+      if (unknownHostException == null) {
+        throw throwable;
+      }
+
+      if (!this.rdsUtils.isRdsDns(hostSpec.getHost()) || !this.rdsUtils.isGreenInstance(hostSpec.getHost())) {
+        throw throwable;
+      }
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java:200]: Does it make sense to check the dns first and fail early before going through potentially 100 exceptions?

File: wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java [hunk 6]
```diff
@@ -157,6 +163,73 @@ public Connection connect(
...
+      } catch (UnknownHostException tmp) {
+        // do nothing
+      }
+
+      if (resolvedAddress != null) {
+        // Green node DNS exists
+        throw throwable;
+      }
+
+      // Green node DNS doesn't exist. Try to replace it with corresponding node name and connect again.
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java:215]:
```suggestion
      // Green node DNS doesn't exist. Try to replace it with the corresponding node name and connect again.
```

File: wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java [hunk 7]
```diff
@@ -157,6 +163,73 @@ public Connection connect(
...
+      // check DNS for such green host name
+      InetAddress resolvedAddress = null;
+      try {
+        resolvedAddress = InetAddress.getByName(hostSpec.getHost());
+      } catch (UnknownHostException tmp) {
+        // do nothing
+      }
+
+      if (resolvedAddress != null) {
+        // Green node DNS exists
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java:211]:
```suggestion
        // Green host DNS exists
```

File: wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java [hunk 8]
```diff
@@ -117,12 +122,71 @@ public Connection connect(
...
         + PropertyUtils.logProperties(PropertyUtils.maskProperties(connectInfo.props), "\nwith properties: \n"));
 
-    Connection conn = this.driver.connect(connectInfo.url, connectInfo.props);
+    Connection conn;
+    try {
+      conn = this.driver.connect(connectInfo.url, connectInfo.props);
+
+    } catch (Throwable throwable) {
+
+      if (!PropertyDefinition.ENABLE_GREEN_NODE_REPLACEMENT.getBoolean(props)) {
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java:137]: What do you think of moving the duplicated code into a util class or helper method?
```java
  public static HostSpec replaceHost(Throwable throwable, RdsUtils rdsUtils, HostSpec hostSpec, Properties props) throws Throwable {
    if (!PropertyDefinition.ENABLE_GREEN_NODE_REPLACEMENT.getBoolean(props)) {
      throw throwable;
    }

    UnknownHostException unknownHostException = null;
    int maxDepth = 100;
    Throwable loopThrowable = throwable;
    while (--maxDepth > 0 && loopThrowable != null) {
      if (loopThrowable instanceof UnknownHostException) {
        unknownHostException = (UnknownHostException) loopThrowable;
        break;
      }
      loopThrowable = loopThrowable.getCause();
    }

    if (unknownHostException == null) {
      throw throwable;
    }

    if (!rdsUtils.isRdsDns(hostSpec.getHost()) || !this.rdsUtils.isGreenInstance(hostSpec.getHost())) {
      throw throwable;
    }

    // check DNS for such green host name
    InetAddress resolvedAddress = null;
    try {
      resolvedAddress = InetAddress.getByName(hostSpec.getHost());
    } catch (UnknownHostException tmp) {
      // do nothing
    }

    if (resolvedAddress != null) {
      // Green host DNS exists
      throw throwable;
    }

    // Green host DNS doesn't exist. Try to replace it with the corresponding hostname and connect again.

    final String fixedHost = rdsUtils.removeGreenInstancePrefix(hostSpec.getHost());
    return new HostSpecBuilder(hostSpec.getHostAvailabilityStrategy())
        .copyFrom(hostSpec)
        .host(fixedHost)
        .build();
  }
```

File: wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java (hunk 5 above)
Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java:198]: I'm not sure. Checking DNS upfront takes time. Checking when exception thrown is handling error. Maybe we will make such change later as part of optimization.

File: wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java (hunk 8 above)
Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java:137]: We can relocate this code to a helper class. The main question whether any other places that require such functionality. I don't see any other place in the driver codebase that may require such checks.

**Comment by davecramer** [comment]
this fixes issue #678, correct ?

**Pull request merged**